File: sources/ShmupWarz/Systems (copy)/ScaleAnimationSystem.fs
namespace ShmupWarz

(**
 * Entitas Generated Systems for ShmupWarz
 *
 *)

open Entitas
open System
open System.Collections.Generic
open UnityEngine

type ScaleAnimationSystem(world:World) =

    let group = world.GetGroup(Matcher.AllOf(Matcher.Scale, Matcher.ScaleAnimation, Matcher.View))

    interface IExecuteSystem with
        member this.Execute() =
            for e in (group.GetEntities()) do
                let scaleAnimation = e.scaleAnimation
                if scaleAnimation.active then
                    let scale = e.scale
                    scale.x <- scale.x + (scaleAnimation.speed * Time.deltaTime)
                    scale.y <- scale.x
                    if scale.x > scaleAnimation.max then
                        scale.x <- scaleAnimation.max
                        scale.y <- scale.x
                        scaleAnimation.active <- false
                    elif scale.x < scaleAnimation.min then
                        scale.x <- scaleAnimation.min
                        scale.y <- scale.x
                        scaleAnimation.active <- false

                    let transform = ((e.view).gameObject:?>GameObject).transform
                    transform.localScale <- new Vector3(scale.x, scale.y, transform.localScale.z)

File: sources/src/systems/ScaleAnimationSystem.ts
module schmupwarz {
  "use strict";

  import Pool = entitas.Pool;
  import Group = entitas.Group;
  import Entity = entitas.Entity;
  import Matcher = entitas.Matcher;
  import Exception = entitas.Exception;
  import TriggerOnEvent = entitas.TriggerOnEvent;
  import IExecuteSystem = entitas.IExecuteSystem;
  import ISetPool = entitas.ISetPool;
  import ScaleAnimation = schmupwarz.ScaleAnimationComponent;
  import Sprite = PIXI.Sprite;

  export class ScaleAnimationSystem implements ISetPool, IExecuteSystem {

    protected pool:Pool;
    protected group:Group;

    public setPool(pool:Pool) {
      this.pool = pool;
      this.group = pool.getGroup(Matcher.allOf(Matcher.ScaleAnimation));
    }
    
    public execute() {
      var delta = bosco.delta;
      var entities = this.group.getEntities();
      for (var i = 0, l = entities.length; i < l; i++) {
        var e:Entity = entities[i];

        var scaleAnimation:ScaleAnimationComponent = e.scaleAnimation;

        if (scaleAnimation.active) {
          var scale = e.sprite.object.scale;

          scale.x += scaleAnimation.speed * delta;
          scale.y = scale.x;

          if (scale.x > scaleAnimation.max) {
            scale.x = scaleAnimation.max;
            scale.y = scale.x;
            scaleAnimation.active = false;
          } else if (scale.x < scaleAnimation.min) {
            scale.x = scaleAnimation.min;
            scale.y = scale.x;
            scaleAnimation.active = false;
          }
        }
      }
    }
  }
}